homeassistant/components/zha/core/device.py

```diff
@@ -0,0 +1,336 @@
+            self.name, self.ieee, SIGNAL_AVAILABLE)
+        self._unsub = async_dispatcher_connect(
+            self.hass,
+            self._available_signal,
+            self.async_initialize
+        )
+
+    @property
+    def name(self):
+        """Return device name."""
+        return self._name
+
+    @property
+    def ieee(self):
+        """Return ieee address for device."""
+        return self._zigpy_device.ieee
+
+    @property
+    def manufacturer(self):
+        """Return ieee address for device."""
+        return self._manufacturer
+
+    @property
+    def model(self):
+        """Return ieee address for device."""
+        return self._model
+
+    @property
+    def nwk(self):
+        """Return nwk for device."""
+        return self._zigpy_device.nwk
+
+    @property
+    def lqi(self):
+        """Return lqi for device."""
+        return self._zigpy_device.lqi
+
+    @property
+    def rssi(self):
+        """Return rssi for device."""
+        return self._zigpy_device.rssi
+
+    @property
+    def last_seen(self):
+        """Return last_seen for device."""
+        return self._zigpy_device.last_seen
+
+    @property
+    def manufacturer_code(self):
+        """Return manufacturer code for device."""
+        # will eventually get this directly from Zigpy
+        return None
+
+    @property
+    def gateway(self):
+        """Return the gateway for this device."""
+        return self._zha_gateway
+
+    @property
+    def cluster_listeners(self):
+        """Return cluster listeners for device."""
+        return self._cluster_listeners.values()
+
+    @property
+    def all_listeners(self):
+        """Return cluster listeners and relay listeners for device."""
+        return self._all_listeners
+
+    @property
+    def cluster_listener_keys(self):
+        """Return cluster listeners for device."""
+        return self._cluster_listeners.keys()
+
+    @property
+    def hass(self):
+        """Return hass."""
+        # Make hass available to use in later processing
+        return self._hass
+
+    @property
+    def available_signal(self):
+        """Signal to use to subscribe to device availability changes."""
+        return self._available_signal
+
+    @property
+    def available(self):
+        """Return True if sensor is available."""
+        return self._available
+
+    @available.setter
+    def available(self, available):
+        """Set sensor availability."""
+        if self._available != available and available:
+            # Update the state the first time the device comes online
+            async_dispatcher_send(
+                self._hass,
+                self._available_signal,
+                False
+            )
+            async_dispatcher_send(
+                self._hass,
+                "{}_{}".format(self._available_signal, 'entity'),
+                True
+            )
+        self._available = available
+
+    @property
+    def device_info(self):
```

Comment:
It’s what the home assistant parts will delegate to. It doesn’t make sense to replicate this everywhere since the home assistant api is exposing “device info” and this object represents the actual device... n many entities can be exposed by a device. Shouldn’t they delegate to this rather than replicate this info in each spot?

Reply:
I think we should be careful not to make this class aware of the home assistant (entity) api. This case is borderline. The rest of the methods are not aware, which is good.

Reply:
I agree. But in this instance I do think it makes sense. What if I remove the HA specific keys and formatting and I return the values that are here in a zha specific dict? Would that be better? 

I was trying to minimize the amount of bridge code in places where it didn’t seem dangerous to blend contexts but I definitely get your point.

Reply:
Yes, returning a dict for later use in `device_info` is fine.

Reply:
Will do